src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeParser.cs

```diff
@@ -433,6 +433,65 @@ private bool ParseMethodCallOrArrayIndex(in SyntaxListBuilder<RazorSyntaxNode> b
                         }
                     }
                 }
+                else if (At(SyntaxKind.Not) && Context.FeatureFlags.AllowNullableForgivenessOperator)
+                {
+                    // C# 8.0 Null forgiveness Operator
+
+                    var next = Lookahead(count: 1);
+                    if (next == null)
+                    {
+                        // Null forgiveness operator at the end of the file, don't include it in the expression.
+                        // We don't allow trailing null forgiveness operators to avoid breaking scenarios such as:
+                        //
+                        // <p>Hello @Person! Good day!</p>
+                        return false;
+                    }
+
+                    if (next.Kind == SyntaxKind.Dot)
+                    {
+                        var nextNext = Lookahead(count: 2);
+                        if (nextNext == null)
+                        {
+                            // End of file after the dot (!.EOF)
+                            return false;
+                        }
+
+                        if (nextNext.Kind == SyntaxKind.Identifier || nextNext.Kind == SyntaxKind.Keyword)
+                        {
+                            // Accept null forgiveness operator followed by a dot (!.)
+                            AcceptAndMoveNext();
+
+                            // Accept the dot
+                            AcceptAndMoveNext();
+                            return true;
+                        }
+
+                        // We're in an odd situation where the user is attempting to use a null-forgiven implicit expression at the
+                        // end of a sentence, i.e.
+                        //
+                        // <p>@Person!.</p>
```

Comment:
Can there ever be generics after the `!.`?

Reply:
Nope, generics aren't allowed in implicit expressions.